File: tests/cli/integration/root-claude-md-preflight.test.ts
/**
 * Integration coverage for R45 strict-abort pre-flight on project-root
 * CLAUDE.md (cw6/T4).
 *
 * The unit tests in tests/state/materialize-section-splice.test.ts pin the
 * preflightRootSplice contract at the materialize-engine level. This file
 * pins the same contract through the spawned CLI:
 *
 *   - Exit code 1 (user error class — see src/cli/exit.ts:exitCodeFor;
 *     MaterializeError → EXIT_USER_ERROR with the "run init" remediation).
 *   - The error message names the file path AND references
 *     `claude-profiles init` so the user has actionable wording.
 *   - Both destinations are byte-identical to their pre-state on abort:
 *     project-root CLAUDE.md AND `.claude/` (R45
 *     atomic-across-destinations).
 *   - No staging artifacts (`.pending/`, `.prior/`, projectRoot `*.tmp`)
 *     remain after the abort.
 *   - The lock is released so a follow-up `init` recovers cleanly.
 *
 * Reverting any of: the preflightRootSplice ordering (must run BEFORE
 * step a), the in-memory snapshot of before/after slices (TOCTOU
 * mitigation), the tmp-cleanup on splice failure, or the lock-release on
 * throw must surface as a failure here.
 *
 * Coverage gap closed: previously the only coverage for the spawn-boundary
 * R45 path was an indirect mention in scenarios.test.ts S11 (which
 * exercises `validate`, not `use`). A future refactor of materialize that
 * wrote bytes to either destination before pre-flight could land green.
 */

import { promises as fs, type Dirent } from "node:fs";
import * as path from "node:path";

import { afterEach, describe, expect, it } from "vitest";

import { merge } from "../../../src/merge/merge.js";
import { resolve } from "../../../src/resolver/resolve.js";
import { materialize } from "../../../src/state/materialize.js";
import { buildStatePaths } from "../../../src/state/paths.js";
import { isRootClaudeMdTmpName } from "../../../src/state/paths.js";
import { makeFixture, type Fixture } from "../../helpers/fixture.js";

import { ensureBuilt, runCli } from "./spawn.js";

let fx: Fixture | undefined;
afterEach(async () => {
  if (fx) await fx.cleanup();
  fx = undefined;
});

/**
 * Build a fixture with a profile `b` that contributes a projectRoot
 * CLAUDE.md (so a `use b` would hit the splice path) plus a `.claude/`
 * file (so step a/b/c would be exercised if pre-flight didn't abort).
 *
 * `start` is the active profile pre-flight. We materialize it directly
 * via the in-process engine so we don't introduce a chicken-and-egg
 * issue with project-root CLAUDE.md handling on the very first run.
 */
async function setupRootContributor(start: "a" = "a"): Promise<Fixture> {
  const f = await makeFixture({
    profiles: {
      a: {
        manifest: { name: "a" },
        files: { "CLAUDE.md": "A\n", "settings.json": '{"v":"a"}' },
      },
      b: {
        manifest: { name: "b" },
        files: { "CLAUDE.md": "B\n", "settings.json": '{"v":"b"}' },
        rootFiles: { "CLAUDE.md": "PROFILE-B-MANAGED-BODY\n" },
      },
    },
  });
  // Materialize `start` (no projectRoot contributor for `a`). The live
  // `.claude/` is now populated; root CLAUDE.md is whatever the test
  // arranges next.
  const plan = await resolve(start, { projectRoot: f.projectRoot });
  const m = await merge(plan);
  await materialize(buildStatePaths(f.projectRoot), plan, m);
  return f;
}

describe("R45 pre-flight — strict abort with malformed root CLAUDE.md (cw6/T4)", () => {
  it("file absent: `use b` aborts; .claude/ byte-identical; exit 1; no leftovers", async () => {
    await ensureBuilt();
    fx = await setupRootContributor();
    const claudeDir = path.join(fx.projectRoot, ".claude");
    const rootClaudeMd = path.join(fx.projectRoot, "CLAUDE.md");

    // Pre-state: capture both destinations byte-by-byte.
    const preClaude = await readTree(claudeDir);
    // No root CLAUDE.md exists.
    await expect(fs.access(rootClaudeMd)).rejects.toMatchObject({ code: "ENOENT" });

    const r = await runCli({ args: ["--cwd", fx.projectRoot, "use", "b"] });

    // Exit code: MaterializeError → EXIT_USER_ERROR (1) per exit.ts.
    expect(r.exitCode).toBe(1);
    // Actionable wording.
    expect(r.stderr).toContain("claude-profiles init");
    expect(r.stderr).toContain("CLAUDE.md");

    // Both destinations untouched (R45 atomic-across-destinations).
    const postClaude = await readTree(claudeDir);
    expect(postClaude).toEqual(preClaude);
    await expect(fs.access(rootClaudeMd)).rejects.toMatchObject({ code: "ENOENT" });

    // No leftover staging.
    const metaDir = path.join(fx.projectRoot, ".claude-profiles", ".meta");
    await expect(
      fs.access(path.join(metaDir, "pending")),
    ).rejects.toMatchObject({ code: "ENOENT" });
    await expect(
      fs.access(path.join(metaDir, "prior")),
    ).rejects.toMatchObject({ code: "ENOENT" });
    await expect(
      fs.access(path.join(metaDir, "lock")),
    ).rejects.toMatchObject({ code: "ENOENT" });
    // No leftover splice tmp at projectRoot.
    const rootEntries = await fs.readdir(fx.projectRoot);
    expect(rootEntries.filter(isRootClaudeMdTmpName)).toHaveLength(0);
  });

  it("file present, markers missing: `use b` aborts; both files byte-identical", async () => {
    await ensureBuilt();
    fx = await setupRootContributor();
    const claudeDir = path.join(fx.projectRoot, ".claude");
    const rootClaudeMd = path.join(fx.projectRoot, "CLAUDE.md");

    // Plain CLAUDE.md without markers — user has not run `init`.
    const original = "# Project README\n\nNo claude-profiles markers here yet.\n";
    await fs.writeFile(rootClaudeMd, original);

    const preClaude = await readTree(claudeDir);

    const r = await runCli({ args: ["--cwd", fx.projectRoot, "use", "b"] });
    expect(r.exitCode).toBe(1);
    expect(r.stderr).toContain("claude-profiles init");

    // R45: BOTH destinations byte-identical.
    expect(await readTree(claudeDir)).toEqual(preClaude);
    expect(await fs.readFile(rootClaudeMd, "utf8")).toBe(original);
  });

  it("file present, markers malformed (lone :begin): aborts; file unchanged", async () => {
    await ensureBuilt();
    fx = await setupRootContributor();
    const rootClaudeMd = path.join(fx.projectRoot, "CLAUDE.md");
    const claudeDir = path.join(fx.projectRoot, ".claude");

    // Lone :begin with no :end. parseMarkers reports "not found" → R45 abort.
    const broken = "intro\n<!-- claude-profiles:v1:begin -->\nstuff but no end\n";
    await fs.writeFile(rootClaudeMd, broken);

    const preClaude = await readTree(claudeDir);

    const r = await runCli({ args: ["--cwd", fx.projectRoot, "use", "b"] });
    expect(r.exitCode).toBe(1);
    expect(await fs.readFile(rootClaudeMd, "utf8")).toBe(broken);
    expect(await readTree(claudeDir)).toEqual(preClaude);
  });

  it("after manual marker recovery, subsequent `use b` succeeds (no leftover state)", async () => {
    // Round-trip: confirm the abort doesn't leave the project in a state
    // that requires anything beyond fixing the markers. Once a well-formed
    // marker pair is in place, the next `use b` lands cleanly — no manual
    // .pending/.prior/.tmp cleanup needed.
    //
    // Note: in this fixture .claude-profiles/ is already populated, so
    // `claude-profiles init` would refuse ("already initialised"). The
    // user's real remediation is to add the marker pair themselves (the
    // error message documents the exact bytes). Here we do that to prove
    // the abort left no other obstacle to recovery.
    await ensureBuilt();
    fx = await setupRootContributor();
    const rootClaudeMd = path.join(fx.projectRoot, "CLAUDE.md");
    const claudeDir = path.join(fx.projectRoot, ".claude");

    // Plain markers-less CLAUDE.md.
    await fs.writeFile(rootClaudeMd, "# Pre-existing notes.\n");

    // First attempt: must abort.
    const r1 = await runCli({ args: ["--cwd", fx.projectRoot, "use", "b"] });
    expect(r1.exitCode).toBe(1);

    // Manually add a valid marker pair (preserving prior content above) —
    // the recovery the error message asks the user to perform.
    await fs.writeFile(
      rootClaudeMd,
      "# Pre-existing notes.\n\n<!-- claude-profiles:v1:begin -->\n<!-- Managed block. -->\n\n<!-- claude-profiles:v1:end -->\n",
    );

    // Now `use b` lands the splice cleanly.
    const r2 = await runCli({ args: ["--cwd", fx.projectRoot, "use", "b"] });
    expect(r2.exitCode).toBe(0);
    expect(r2.stdout).toContain("Switched to b");
    // The user's prose above the markers is preserved byte-for-byte.
    const final = await fs.readFile(rootClaudeMd, "utf8");
    expect(final.startsWith("# Pre-existing notes.")).toBe(true);
    expect(final).toContain("PROFILE-B-MANAGED-BODY");
    // .claude/ landed b's content.
    expect(
      await fs.readFile(path.join(claudeDir, "CLAUDE.md"), "utf8"),
    ).toBe("B\n");
  });

  it("multiple managed blocks (currently rejected): aborts; file unchanged", async () => {
    // Per docs/specs §12.3 multiple managed blocks are reserved for future
    // versions and v1 implementations may treat as malformed. Exercise that
    // path so a future relaxation needs an explicit test update.
    await ensureBuilt();
    fx = await setupRootContributor();
    const rootClaudeMd = path.join(fx.projectRoot, "CLAUDE.md");
    const claudeDir = path.join(fx.projectRoot, ".claude");

    const dual =
      "intro\n" +
      "<!-- claude-profiles:v1:begin -->\nblock 1\n<!-- claude-profiles:v1:end -->\n" +
      "between\n" +
      "<!-- claude-profiles:v1:begin -->\nblock 2\n<!-- claude-profiles:v1:end -->\n" +
      "tail\n";
    await fs.writeFile(rootClaudeMd, dual);

    const preClaude = await readTree(claudeDir);

    const r = await runCli({ args: ["--cwd", fx.projectRoot, "use", "b"] });
    expect(r.exitCode).toBe(1);
    expect(await fs.readFile(rootClaudeMd, "utf8")).toBe(dual);
    expect(await readTree(claudeDir)).toEqual(preClaude);
  });
});

/**
 * Recursively read every regular file under `root` keyed by relative posix
 * path → utf8 content. Returns {} for a missing root so pre/post comparisons
 * stay uniform.
 */
async function readTree(root: string): Promise<Record<string, string>> {
  const out: Record<string, string> = {};
  async function walk(rel: string): Promise<void> {
    const here = rel === "" ? root : path.join(root, rel);
    let entries: Dirent[];
    try {
      entries = (await fs.readdir(here, { withFileTypes: true })) as Dirent[];
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code === "ENOENT") return;
      throw err;
    }
    for (const e of entries) {
      const name = String(e.name);
      const childRel = rel === "" ? name : `${rel}/${name}`;
      const childAbs = path.join(here, name);
      if (e.isDirectory()) {
        await walk(childRel);
      } else if (e.isFile()) {
        out[childRel] = await fs.readFile(childAbs, "utf8");
      }
    }
  }
  await walk("");
  return out;
}

File: go/tests/integration/root_claude_md_preflight_test.go
package integration_test

import (
	"errors"
	"io/fs"
	"os"
	"path/filepath"
	"reflect"
	"regexp"
	"strings"
	"testing"

	"github.com/htxryan/c3p/tests/integration/helpers"
)

// IV translation of tests/cli/integration/root-claude-md-preflight.test.ts.
// Pins R45 strict-abort pre-flight on project-root CLAUDE.md (cw6/T4)
// through the spawned Go CLI:
//   - Exit code 1 (user error class).
//   - Error wording names CLAUDE.md and `c3p init`.
//   - Both destinations (project-root CLAUDE.md and .claude/) byte-identical
//     to pre-state on abort (R45 atomic-across-destinations).
//   - No leftover .pending / .prior / projectRoot CLAUDE.md.*.tmp staging.
//
// The lock file is a record of the last lock holder in the Go bin and is
// expected to remain (unlike the TS bin); we don't pin its absence here.

// rootClaudeMdTmpPattern mirrors state.IsRootClaudeMdTmpName; we duplicate
// the regex (rather than import the internal package) per IV constraint:
// tests/integration must talk to the bin only.
var rootClaudeMdTmpPattern = regexp.MustCompile(`^CLAUDE\.md\.\d+\.\d+-[a-z0-9]+\.tmp$`)

// setupRootContributorFx builds a fixture with profile `a` (no projectRoot
// contributor) and profile `b` (root + .claude/ contributors), then
// materializes `a` via the CLI so .claude/ has a defined pre-state.
func setupRootContributorFx(t *testing.T) *helpers.Fixture {
	t.Helper()
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{
		Profiles: map[string]helpers.ProfileSpec{
			"a": {
				Manifest: map[string]any{"name": "a"},
				Files:    map[string]string{"CLAUDE.md": "A\n", "settings.json": `{"v":"a"}`},
			},
			"b": {
				Manifest:  map[string]any{"name": "b"},
				Files:     map[string]string{"CLAUDE.md": "B\n", "settings.json": `{"v":"b"}`},
				RootFiles: map[string]string{"CLAUDE.md": "PROFILE-B-MANAGED-BODY\n"},
			},
		},
	})
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"--cwd", fx.ProjectRoot, "use", "a"}})
	if r.ExitCode != 0 {
		t.Fatalf("setup use a: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	return fx
}

// readTree mirrors the TS helper: relative posix path -> utf8 content.
// Returns empty map for missing root.
func readTree(t *testing.T, root string) map[string]string {
	t.Helper()
	out := map[string]string{}
	err := filepath.WalkDir(root, func(p string, d fs.DirEntry, err error) error {
		if err != nil {
			if errors.Is(err, fs.ErrNotExist) {
				return nil
			}
			return err
		}
		if d.IsDir() {
			return nil
		}
		rel, err := filepath.Rel(root, p)
		if err != nil {
			return err
		}
		b, err := os.ReadFile(p)
		if err != nil {
			return err
		}
		out[filepath.ToSlash(rel)] = string(b)
		return nil
	})
	if err != nil && !errors.Is(err, fs.ErrNotExist) {
		t.Fatalf("readTree %q: %v", root, err)
	}
	return out
}

func assertNotExist(t *testing.T, p string) {
	t.Helper()
	if _, err := os.Stat(p); err == nil {
		t.Errorf("expected %q absent, but it exists", p)
	} else if !errors.Is(err, fs.ErrNotExist) {
		t.Errorf("stat %q: unexpected err %v", p, err)
	}
}

// TestRootClaudeMdPreflight_FileAbsent — `use b` aborts when projectRoot
// CLAUDE.md is missing; .claude/ byte-identical; exit 1; no .pending/.prior;
// no leftover CLAUDE.md.*.tmp files in projectRoot.
func TestRootClaudeMdPreflight_FileAbsent(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupRootContributorFx(t)
	claudeDir := filepath.Join(fx.ProjectRoot, ".claude")
	rootClaudeMd := filepath.Join(fx.ProjectRoot, "CLAUDE.md")

	preClaude := readTree(t, claudeDir)
	assertNotExist(t, rootClaudeMd)

	r := mustRun(t, helpers.SpawnOptions{Args: []string{"--cwd", fx.ProjectRoot, "use", "b"}})

	if r.ExitCode != 1 {
		t.Fatalf("use b (no root): want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stderr, "c3p init") {
		t.Errorf("stderr missing 'c3p init': %q", r.Stderr)
	}
	if !strings.Contains(r.Stderr, "CLAUDE.md") {
		t.Errorf("stderr missing 'CLAUDE.md': %q", r.Stderr)
	}

	// Both destinations untouched.
	post := readTree(t, claudeDir)
	if !reflect.DeepEqual(post, preClaude) {
		t.Errorf("claude/ tree changed: pre=%v post=%v", preClaude, post)
	}
	assertNotExist(t, rootClaudeMd)

	// No leftover staging.
	metaDir := filepath.Join(fx.ProjectRoot, ".claude-profiles", ".meta")
	assertNotExist(t, filepath.Join(metaDir, "pending"))
	assertNotExist(t, filepath.Join(metaDir, "prior"))

	// No leftover splice tmp at projectRoot.
	entries, err := os.ReadDir(fx.ProjectRoot)
	if err != nil {
		t.Fatalf("readdir projectRoot: %v", err)
	}
	for _, e := range entries {
		if rootClaudeMdTmpPattern.MatchString(e.Name()) {
			t.Errorf("leftover splice tmp: %q", e.Name())
		}
	}
}

// TestRootClaudeMdPreflight_FilePresentMarkersMissing — plain CLAUDE.md
// without c3p markers triggers strict abort; both destinations byte-identical.
func TestRootClaudeMdPreflight_FilePresentMarkersMissing(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupRootContributorFx(t)
	claudeDir := filepath.Join(fx.ProjectRoot, ".claude")
	rootClaudeMd := filepath.Join(fx.ProjectRoot, "CLAUDE.md")

	original := "# Project README\n\nNo c3p markers here yet.\n"
	if err := os.WriteFile(rootClaudeMd, []byte(original), 0o644); err != nil {
		t.Fatalf("seed CLAUDE.md: %v", err)
	}
	preClaude := readTree(t, claudeDir)

	r := mustRun(t, helpers.SpawnOptions{Args: []string{"--cwd", fx.ProjectRoot, "use", "b"}})
	if r.ExitCode != 1 {
		t.Fatalf("use b: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stderr, "c3p init") {
		t.Errorf("stderr missing 'c3p init': %q", r.Stderr)
	}
	if got := readTree(t, claudeDir); !reflect.DeepEqual(got, preClaude) {
		t.Errorf("claude/ tree changed")
	}
	got, err := os.ReadFile(rootClaudeMd)
	if err != nil {
		t.Fatalf("read CLAUDE.md: %v", err)
	}
	if string(got) != original {
		t.Errorf("CLAUDE.md changed: got %q want %q", string(got), original)
	}
}

// TestRootClaudeMdPreflight_MalformedLoneBegin — lone :begin without :end
// is treated as malformed → R45 abort; file unchanged.
func TestRootClaudeMdPreflight_MalformedLoneBegin(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupRootContributorFx(t)
	rootClaudeMd := filepath.Join(fx.ProjectRoot, "CLAUDE.md")
	claudeDir := filepath.Join(fx.ProjectRoot, ".claude")

	broken := "intro\n<!-- c3p:v1:begin -->\nstuff but no end\n"
	if err := os.WriteFile(rootClaudeMd, []byte(broken), 0o644); err != nil {
		t.Fatalf("seed: %v", err)
	}
	preClaude := readTree(t, claudeDir)

	r := mustRun(t, helpers.SpawnOptions{Args: []string{"--cwd", fx.ProjectRoot, "use", "b"}})
	if r.ExitCode != 1 {
		t.Fatalf("use b malformed: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	got, err := os.ReadFile(rootClaudeMd)
	if err != nil {
		t.Fatalf("read CLAUDE.md: %v", err)
	}
	if string(got) != broken {
		t.Errorf("CLAUDE.md mutated on abort: got %q", string(got))
	}
	if post := readTree(t, claudeDir); !reflect.DeepEqual(post, preClaude) {
		t.Errorf("claude/ tree changed")
	}
}

// TestRootClaudeMdPreflight_RecoversAfterMarkerFix — once the user adds a
// well-formed marker pair, the next `use b` lands the splice cleanly with
// no manual cleanup of staging dirs needed. Note: the Go bin emits
// "switched to b" (lowercase verb), not "Switched to b" as TS does.
func TestRootClaudeMdPreflight_RecoversAfterMarkerFix(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupRootContributorFx(t)
	rootClaudeMd := filepath.Join(fx.ProjectRoot, "CLAUDE.md")
	claudeDir := filepath.Join(fx.ProjectRoot, ".claude")

	if err := os.WriteFile(rootClaudeMd, []byte("# Pre-existing notes.\n"), 0o644); err != nil {
		t.Fatalf("seed: %v", err)
	}

	r1 := mustRun(t, helpers.SpawnOptions{Args: []string{"--cwd", fx.ProjectRoot, "use", "b"}})
	if r1.ExitCode != 1 {
		t.Fatalf("first use b: want 1, got %d (stderr=%q)", r1.ExitCode, r1.Stderr)
	}

	fixed := "# Pre-existing notes.\n\n<!-- c3p:v1:begin -->\n<!-- Managed block. -->\n\n<!-- c3p:v1:end -->\n"
	if err := os.WriteFile(rootClaudeMd, []byte(fixed), 0o644); err != nil {
		t.Fatalf("write fixed: %v", err)
	}

	r2 := mustRun(t, helpers.SpawnOptions{Args: []string{"--cwd", fx.ProjectRoot, "use", "b"}})
	if r2.ExitCode != 0 {
		t.Fatalf("retry use b: want 0, got %d (stderr=%q)", r2.ExitCode, r2.Stderr)
	}
	if !strings.Contains(strings.ToLower(r2.Stdout), "switched to b") {
		t.Errorf("stdout missing 'switched to b': %q", r2.Stdout)
	}

	final, err := os.ReadFile(rootClaudeMd)
	if err != nil {
		t.Fatalf("read final: %v", err)
	}
	if !strings.HasPrefix(string(final), "# Pre-existing notes.") {
		t.Errorf("user prose dropped: %q", string(final))
	}
	if !strings.Contains(string(final), "PROFILE-B-MANAGED-BODY") {
		t.Errorf("managed body not spliced: %q", string(final))
	}
	got, err := os.ReadFile(filepath.Join(claudeDir, "CLAUDE.md"))
	if err != nil {
		t.Fatalf("read .claude/CLAUDE.md: %v", err)
	}
	if string(got) != "B\n" {
		t.Errorf(".claude/CLAUDE.md = %q, want %q", string(got), "B\n")
	}
}

// TestRootClaudeMdPreflight_MultipleManagedBlocks — per spec §12.3 multiple
// managed blocks are reserved; v1 treats as malformed → abort, file unchanged.
func TestRootClaudeMdPreflight_MultipleManagedBlocks(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := setupRootContributorFx(t)
	rootClaudeMd := filepath.Join(fx.ProjectRoot, "CLAUDE.md")
	claudeDir := filepath.Join(fx.ProjectRoot, ".claude")

	dual := "intro\n" +
		"<!-- c3p:v1:begin -->\nblock 1\n<!-- c3p:v1:end -->\n" +
		"between\n" +
		"<!-- c3p:v1:begin -->\nblock 2\n<!-- c3p:v1:end -->\n" +
		"tail\n"
	if err := os.WriteFile(rootClaudeMd, []byte(dual), 0o644); err != nil {
		t.Fatalf("seed dual: %v", err)
	}
	preClaude := readTree(t, claudeDir)

	r := mustRun(t, helpers.SpawnOptions{Args: []string{"--cwd", fx.ProjectRoot, "use", "b"}})
	if r.ExitCode != 1 {
		t.Fatalf("use b dual: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	got, err := os.ReadFile(rootClaudeMd)
	if err != nil {
		t.Fatalf("read: %v", err)
	}
	if string(got) != dual {
		t.Errorf("CLAUDE.md mutated: got %q", string(got))
	}
	if post := readTree(t, claudeDir); !reflect.DeepEqual(post, preClaude) {
		t.Errorf("claude/ tree changed")
	}
}
